apps/site/assets/ts/helpers/departureInfo.tsx

```diff
@@ -123,11 +135,54 @@ const isCommuterRail = (departureInfo: DepartureInfo): boolean => {
   return departureInfo?.routeMode === "commuter_rail";
 };
 
+const DefaultWrapper: React.FunctionComponent<unknown> = ({ children }) => (
+  <li>{children}</li>
+);
+/**
+ * From a list of `DepartureInfo[]`, generate a list of `<DisplayTime />`.
+ * Each `<DisplayTime />` will recieve the `isCR` and `targetDate` props.
+ *
+ * This function can optionally
+ * - wrap each `<DisplayTime />` in a specified wrapper element, `WrapperEl`
+ * - truncate the list to a desired `listLength` value
+ */
+const departuresListFromInfos = (
+  departureInfos: DepartureInfo[],
+  isCR: boolean,
+  targetDate?: Date,
+  listLength?: number,
+  WrapperEl: typeof DefaultWrapper = DefaultWrapper
+): React.ReactElement[] => {
+  // optional cutoff time, before which we won't show schedules.
+  // just used with subway for now.
+  const predictionTimeCutoff = chain(departureInfos)
+    .filter(d => d.routeMode === SUBWAY)
+    .maxBy("prediction.time")
+    .value()?.prediction?.time;
+
+  return chain(departureInfos)
+    .omitBy(
+      ({ prediction, schedule }) =>
+        predictionTimeCutoff &&
+        !prediction &&
+        schedule &&
+        schedule.time <= predictionTimeCutoff
+    )
+    .map(d => (
+      <WrapperEl key={getInfoKey(d)}>
+        <DisplayTime departure={d} isCR={isCR} targetDate={targetDate} />
+      </WrapperEl>
+    ))
+    .slice(0, listLength)
+    .value();
```

Comment:
Why is this complex `chain` logic needed over just using the `prediction` from the `departureInfos`?  Based of the ticket it seems like the goal is just to show predictions if they exist, which I thought we were doing already.
The `departureInfoToTime` function returns the prediction time in favor of the schedule time.
Maybe I am having a hard time following this logic.

Reply:
@kotva006 Here's a non-chained translation, if it's helpful 🤔 

```javascript
const departuresListWithoutIntermingledSchedules = omitBy(departureInfos, departure => {
  if (predictionTimeCutoff) {
    // omit schedule-only departures that are before the latest prediction time
    !prediction && schedule && schedule.time <= predictionTimeCutoff;
  }
  return false;
});
const departureElements = departuresListWithoutIntermingledSchedules.map(d => /* render as <DisplayTime /> and wrap in wrapper element */)
return departureElements.slice(0, listLength);
```